tempodb/encoding/vparquet4/schema.go

```diff
@@ -0,0 +1,963 @@
+			intValues := make([]v1.AnyValue_DoubleValue, len(values))
+			for i, v := range attr.ValueDouble {
+				n := &intValues[i]
+				n.DoubleValue = v
+				values[i] = &anyValues[i]
+				values[i].Value = n
+			}
+
+			protoVal.Value = &v1.AnyValue_ArrayValue{ArrayValue: &v1.ArrayValue{Values: values}}
+		case attrTypeBoolArray:
+			values := make([]*v1.AnyValue, len(attr.ValueBool))
+
+			anyValues := make([]v1.AnyValue, len(values))
+			intValues := make([]v1.AnyValue_BoolValue, len(values))
+			for i, v := range attr.ValueBool {
+				n := &intValues[i]
+				n.BoolValue = v
+				values[i] = &anyValues[i]
+				values[i].Value = n
+			}
+
+			protoVal.Value = &v1.AnyValue_ArrayValue{ArrayValue: &v1.ArrayValue{Values: values}}
+		case attrTypeNotSupported:
+			if attr.ValueDropped == "" || !includeDroppedAttr {
+				continue
+			}
+			_ = jsonpb.Unmarshal(bytes.NewBufferString(attr.ValueDropped), &protoVal)
+			counter.addDroppedAttr(-1)
+		default:
+			continue
+		}
+
+		protoAttrs = append(protoAttrs, &v1.KeyValue{
+			Key:   attr.Key,
+			Value: &protoVal,
+		})
+	}
+
+	return protoAttrs
+}
+
+func parquetToProtoInstrumentationScope(parquetScope *InstrumentationScope) *v1.InstrumentationScope {
+	scope := v1.InstrumentationScope{
+		Name:    parquetScope.Name,
+		Version: parquetScope.Version,
+	}
+	// TODO: handle attributes correctly once they are added to the proto
+	return &scope
+}
+
+func parquetToProtoLinks(parquetLinks []Link) []*v1_trace.Span_Link {
+	var protoLinks []*v1_trace.Span_Link
+
+	if len(parquetLinks) > 0 {
+		protoLinks = make([]*v1_trace.Span_Link, 0, len(parquetLinks))
+		for _, l := range parquetLinks {
+			protoLink := &v1_trace.Span_Link{
+				TraceId:                l.TraceID,
+				SpanId:                 l.SpanID,
+				TraceState:             l.TraceState,
+				DroppedAttributesCount: uint32(l.DroppedAttributesCount),
+				Attributes:             nil,
+			}
+
+			if len(l.Attrs) > 0 {
+				protoLink.Attributes = parquetToProtoAttrs(l.Attrs, &l, false)
+			}
+
+			protoLinks = append(protoLinks, protoLink)
+		}
+	}
+
+	return protoLinks
+}
+
+func parquetToProtoEvents(parquetEvents []Event, spanStartTimeNano uint64) []*v1_trace.Span_Event {
+	var protoEvents []*v1_trace.Span_Event
+
+	if len(parquetEvents) > 0 {
+		protoEvents = make([]*v1_trace.Span_Event, 0, len(parquetEvents))
+
+		for _, e := range parquetEvents {
+
+			protoEvent := &v1_trace.Span_Event{
+				TimeUnixNano:           e.TimeSinceStartNano + spanStartTimeNano,
+				Name:                   e.Name,
+				Attributes:             nil,
+				DroppedAttributesCount: uint32(e.DroppedAttributesCount),
+			}
+
+			if len(e.Attrs) > 0 {
+				protoEvent.Attributes = parquetToProtoAttrs(e.Attrs, &e, false)
+			}
+
+			protoEvents = append(protoEvents, protoEvent)
+		}
+	}
+
+	return protoEvents
+}
+
+func parquetTraceToTempopbTrace(meta *backend.BlockMeta, parquetTrace *Trace, includeDroppedAttr bool) *tempopb.Trace {
```

Comment:
it's not clear to me why `includeDroppedAttr` is a boolean. would we not want to always return the count of dropped attrs?

Reply:
This is not quite what `includeDroppedAttr` does
* If `includeDroppedAttr` is `true` the converted `tempopb.Trace` will contain all attributes that are dropped by Tempo as if they weren't dropped and will also not show the incremented dropped attribute count
* If it is `false` dropped attributes will not appear in the converted trace and the dropped attribute count will be incremented by one for every dropped attribute. On most places this function is called with `false` i.e. from the outside dropped attributes will appear as dropped: They are not searchable with TraceQL, can't be selected, and do not appear in traces found via trace by ID

Reply:
let's back up and discuss dropped attributes handling. in my opinion we should simplify this logic by:

- removing the `includeDroppedAttr` boolean and always rendering dropped attributes. we can just return the json string.
- removing the logic that internally increments/decrements dropped attributes and just faithfully storing whatever the trace pipeline sends us.

wdyt?

Reply:
Sounds good. I added a commit to change this behavior. Attributes with unsupported values are no longer treated as dropped values:
* Changed the column name from `ValueDropped` to `ValueUnsupported`
* Removed the logic to increment `DroppedAttributesCount`
* Removed the parameter `includeDroppedAttr` and the associated logic.